File: clients/python/jev_fabric.py
"""A thin client for `jev-fabric -- serve`, the JSONL session protocol.

One file, standard library only, Python 3.9+. Copy it next to your code, or
import it from ``~/.local/share/jev-fabric/current/clients/python``.

The client starts one ``serve`` child, writes one request line per call and
reads the matching response line. Budgets, deadlines, credentials and Jev
validation all live in the executable; this module only frames JSON.

    from jev_fabric import Fabric

    with Fabric(max_evaluations=10) as fabric:
        job = fabric.start(["/bin/sh", "-c", "npm run dev"])
        fabric.watch(job, "ready", timeout_ms=30000)
        answer = fabric.jev(request)
"""

from __future__ import annotations

import itertools
import json
import os
import subprocess
import threading
from typing import Any, Dict, List, Mapping, Optional, Sequence

__all__ = ["Fabric", "FabricError", "PROTOCOL"]

PROTOCOL = 1

Json = Dict[str, Any]


class FabricError(Exception):
    """A request the session refused or could not complete.

    ``code`` follows the CLI's exit codes: 2 for a malformed request, 22 for a
    rejected value, 124 for an expired deadline, 1 otherwise.
    """

    def __init__(self, code: int, message: str, op: Optional[str] = None) -> None:
        super().__init__(f"{op}: {message}" if op else message)
        self.code = code
        self.message = message
        self.op = op


class Fabric:
    """One ``jev-fabric -- serve`` session.

    ``timeout_ms`` bounds the whole session (default: the CLI work default, one
    hour). ``max_evaluations`` and ``max_tokens`` bound Jev calls for the whole
    session (defaults: 1 evaluation, 100000 reported tokens, as for ``jev``).
    ``binary`` defaults to ``$JEV_FABRIC_BIN`` or ``jev-fabric`` on PATH.

    Calls run one at a time; a lock makes the object safe to share between
    threads, but requests never overlap.
    """

    def __init__(
        self,
        *,
        binary: Optional[str] = None,
        timeout_ms: Optional[int] = None,
        max_evaluations: Optional[int] = None,
        max_tokens: Optional[int] = None,
        env: Optional[Mapping[str, str]] = None,
        cwd: Optional[str] = None,
    ) -> None:
        executable = binary or os.environ.get("JEV_FABRIC_BIN") or "jev-fabric"
        argv = [executable, "--", "serve"]
        if timeout_ms is not None:
            argv += ["--timeout-ms", str(timeout_ms)]
        if max_tokens is not None and max_evaluations is None:
            max_evaluations = 1
        if max_evaluations is not None:
            argv.append(str(max_evaluations))
        if max_tokens is not None:
            argv.append(str(max_tokens))
        self._process = subprocess.Popen(
            argv,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            env=None if env is None else dict(env),
            cwd=cwd,
        )
        self._ids = itertools.count(1)
        self._lock = threading.Lock()
        banner = self._read_line()
        ready = banner.get("ready")
        if not isinstance(ready, dict) or ready.get("protocol") != PROTOCOL:
            self.close()
            raise FabricError(1, f"unsupported serve protocol: {banner!r}")
        self.ready: Json = ready

    # -- processes -----------------------------------------------------------

    def exec(
        self,
        argv: Sequence[str],
        *,
        stdin: Optional[str] = None,
        timeout_ms: Optional[int] = None,
    ) -> Json:
        """Run literal argv to completion and return its bounded receipt.

        A nonzero exit is a receipt with ``state: "failed"``, not an error.
        """
        return self._call("exec", argv=list(argv), stdin=stdin, timeoutMs=timeout_ms)

    def start(self, argv: Sequence[str], *, timeout_ms: Optional[int] = None) -> str:
        """Start a detached job that outlives this session; returns its id."""
        return self._call("start", argv=list(argv), timeoutMs=timeout_ms)["id"]

    def status(self, job: str) -> Json:
        return self._call("status", job=job)

    def events(self, job: str, *, after: int = 0) -> List[Json]:
        """Retained events with a sequence above ``after`` (a bounded snapshot)."""
        return self._call("events", job=job, after=after)

    def wait(self, job: str, *, timeout_ms: Optional[int] = None) -> Json:
        """The final receipt, or the running state once ``timeout_ms`` passes."""
        return self._call("wait", job=job, timeoutMs=timeout_ms)

    def stop(self, job: str) -> Json:
        return self._call("stop", job=job)

    def watch(self, job: str, literal: str, *, timeout_ms: Optional[int] = None) -> List[Json]:
        """Live output lines containing ``literal``, as monitor records."""
        return self._call("watch", job=job, literal=literal, timeoutMs=timeout_ms)

    # -- Jev -----------------------------------------------------------------

    def validate(self, request: Json) -> Json:
        """Strictly validate a Jev request offline, without credentials."""
        return self._call("validate", request=request)

    def jev(self, request: Json, *, timeout_ms: Optional[int] = None) -> Json:
        """One explicit, billed evaluation against the session budget."""
        return self._call("jev", request=request, timeoutMs=timeout_ms)

    # -- session -------------------------------------------------------------

    def close(self, timeout: float = 10.0) -> int:
        """End the session by closing its input; returns the exit code."""
        process = self._process
        if process.stdin and not process.stdin.closed:
            try:
                process.stdin.close()
            except BrokenPipeError:
                pass
        try:
            return process.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            process.kill()
            return process.wait()

    def __enter__(self) -> "Fabric":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    # -- framing -------------------------------------------------------------

    def _call(self, op: str, **fields: Any) -> Any:
        with self._lock:
            request_id = next(self._ids)
            request = {"id": request_id, "op": op}
            request.update((k, v) for k, v in fields.items() if v is not None)
            line = json.dumps(request, ensure_ascii=False, separators=(",", ":"))
            try:
                assert self._process.stdin is not None
                self._process.stdin.write(line.encode("utf-8") + b"\n")
                self._process.stdin.flush()
            except (BrokenPipeError, ValueError):
                raise self._ended(op) from None
            response = self._read_line(op)
        if response.get("id") != request_id:
            raise FabricError(1, f"response id {response.get('id')!r} does not match {request_id}", op)
        if response.get("ok"):
            return response["result"]
        error = response.get("error") or {}
        raise FabricError(int(error.get("code", 1)), str(error.get("message", "request failed")), op)

    def _read_line(self, op: Optional[str] = None) -> Json:
        assert self._process.stdout is not None
        line = self._process.stdout.readline()
        if not line:
            raise self._ended(op)
        return json.loads(line)

    def _ended(self, op: Optional[str]) -> FabricError:
        code = self._process.wait()
        stderr = self._process.stderr.read().decode("utf-8", "replace").strip() if self._process.stderr else ""
        return FabricError(code or 1, stderr or f"serve exited with code {code}", op)

File: clients/typescript/jev-fabric.ts
/**
 * A thin client for `jev-fabric -- serve`, the JSONL session protocol.
 *
 * One file, Node built-ins only, erasable TypeScript: Bun, Deno, Node 22.6+
 * (type stripping) or any bundler.
 * Copy it next to your code, or import it from
 * `~/.local/share/jev-fabric/current/clients/typescript`.
 *
 * The client starts one `serve` child, writes one request line per call and
 * resolves each promise with the matching response line. Budgets, deadlines,
 * credentials and Jev validation all live in the executable; this module only
 * frames JSON.
 *
 * ```ts
 * import { Fabric } from './jev-fabric.ts';
 *
 * const fabric = await Fabric.open({ maxEvaluations: 10 });
 * const job = await fabric.start(['/bin/sh', '-c', 'npm run dev']);
 * await fabric.watch(job, 'ready', { timeoutMs: 30000 });
 * const answer = await fabric.jev(request);
 * await fabric.close();
 * ```
 */
import { spawn, type ChildProcessWithoutNullStreams } from 'node:child_process';
import { createInterface } from 'node:readline';

export const PROTOCOL = 1;

export interface FabricOptions {
  /** Defaults to `$JEV_FABRIC_BIN` or `jev-fabric` on PATH. */
  binary?: string;
  /** Bounds the whole session; defaults to the CLI work default (one hour). */
  timeoutMs?: number;
  /** Jev evaluations for the whole session; defaults to 1, as for `jev`. */
  maxEvaluations?: number;
  /** Reported Jev tokens for the whole session; defaults to 100000. */
  maxTokens?: number;
  env?: NodeJS.ProcessEnv;
  cwd?: string;
}

export interface Ready {
  protocol: number;
  version: string;
  timeoutMs: number;
  maxEvaluations: number;
  maxTokens: number;
}

export interface Receipt {
  schemaVersion: number;
  state: 'exited' | 'failed' | 'timed_out' | 'cancelled';
  exitCode: number | null;
  timedOut: boolean;
  cancelled: boolean;
  stdout: string;
  stderr: string;
  truncated: { stdout: boolean; stderr: boolean };
}

export type JobState =
  | (Receipt & { id: string; spoolLimitBytes: number })
  | { schemaVersion: number; id: string; state: 'running'; spoolLimitBytes: number }
  | { schemaVersion: number; id: string; state: 'failed'; exitCode: null; error: string; spoolLimitBytes: number };

export interface JobEvent {
  sequence: number;
  type: string;
  data: unknown;
}

export interface MonitorRecord {
  type: 'monitor.batch' | 'monitor.loss' | 'monitor.end';
  [field: string]: unknown;
}

export type JevRequest = Record<string, unknown>;
export type JevAnswer = {
  model?: string;
  answers: Record<string, Record<string, unknown>>;
  usage?: { input_tokens: number; output_tokens: number };
};

/**
 * A request the session refused or could not complete. `code` follows the
 * CLI's exit codes: 2 for a malformed request, 22 for a rejected value, 124 for
 * an expired deadline, 1 otherwise.
 */
export class FabricError extends Error {
  readonly code: number;
  readonly op?: string;

  constructor(code: number, message: string, op?: string) {
    super(op ? `${op}: ${message}` : message);
    this.name = 'FabricError';
    this.code = code;
    this.op = op;
  }
}

interface Pending {
  op: string;
  resolve(value: any): void;
  reject(error: FabricError): void;
}

/**
 * One `jev-fabric -- serve` session. Requests may be issued concurrently;
 * the session answers them one at a time, in the order they were sent.
 */
export class Fabric {
  readonly ready: Ready;
  private readonly child: ChildProcessWithoutNullStreams;
  private readonly pending = new Map<number, Pending>();
  private nextId = 1;
  private readonly exited: Promise<number | null>;
  private readonly log: { text: string };
  private ended?: FabricError;

  private constructor(
    child: ChildProcessWithoutNullStreams,
    ready: Ready,
    lines: AsyncIterator<string>,
    exited: Promise<number | null>,
    log: { text: string },
  ) {
    this.child = child;
    this.ready = ready;
    this.exited = exited;
    this.log = log;
    void this.pump(lines);
  }

  static async open(options: FabricOptions = {}): Promise<Fabric> {
    const binary = options.binary ?? process.env.JEV_FABRIC_BIN ?? 'jev-fabric';
    const argv = ['--', 'serve'];
    if (options.timeoutMs !== undefined) argv.push('--timeout-ms', String(options.timeoutMs));
    const maxEvaluations = options.maxEvaluations ?? (options.maxTokens !== undefined ? 1 : undefined);
    if (maxEvaluations !== undefined) argv.push(String(maxEvaluations));
    if (options.maxTokens !== undefined) argv.push(String(options.maxTokens));
    const child = spawn(binary, argv, { env: options.env, cwd: options.cwd, stdio: 'pipe' });
    // serve writes to stderr only when it exits abnormally; keep it for the error.
    const log = { text: '' };
    child.stderr.setEncoding('utf8');
    child.stderr.on('data', chunk => (log.text += chunk));
    const exited = new Promise<number | null>(done => child.once('close', code => done(code)));
    await new Promise<void>((resolve, reject) => {
      child.once('spawn', resolve);
      child.once('error', error => reject(new FabricError(1, `could not start ${binary}: ${error.message}`)));
    });
    const lines = createInterface({ input: child.stdout, crlfDelay: Infinity })[Symbol.asyncIterator]();
    const first = await lines.next();
    if (first.done) {
      const code = await exited;
      throw new FabricError(code || 1, log.text.trim() || `serve exited with code ${code}`);
    }
    const banner = JSON.parse(first.value);
    if (banner?.ready?.protocol !== PROTOCOL) {
      child.stdin.end();
      throw new FabricError(1, `unsupported serve protocol: ${first.value}`);
    }
    return new Fabric(child, banner.ready, lines, exited, log);
  }

  // -- processes -------------------------------------------------------------

  /** Runs literal argv to completion. A nonzero exit is a receipt, not an error. */
  exec(argv: string[], options: { stdin?: string; timeoutMs?: number } = {}): Promise<Receipt> {
    return this.call('exec', { argv, stdin: options.stdin, timeoutMs: options.timeoutMs });
  }

  /** Starts a detached job that outlives this session; resolves to its id. */
  async start(argv: string[], options: { timeoutMs?: number } = {}): Promise<string> {
    const started = await this.call<{ id: string }>('start', { argv, timeoutMs: options.timeoutMs });
    return started.id;
  }

  status(job: string): Promise<JobState> {
    return this.call('status', { job });
  }

  /** Retained events with a sequence above `after` (a bounded snapshot). */
  events(job: string, options: { after?: number } = {}): Promise<JobEvent[]> {
    return this.call('events', { job, after: options.after });
  }

  /** The final receipt, or the running state once `timeoutMs` passes. */
  wait(job: string, options: { timeoutMs?: number } = {}): Promise<JobState> {
    return this.call('wait', { job, timeoutMs: options.timeoutMs });
  }

  stop(job: string): Promise<JobState> {
    return this.call('stop', { job });
  }

  /** Live output lines containing `literal`, as monitor records. */
  watch(job: string, literal: string, options: { timeoutMs?: number } = {}): Promise<MonitorRecord[]> {
    return this.call('watch', { job, literal, timeoutMs: options.timeoutMs });
  }

  // -- Jev -------------------------------------------------------------------

  /** Strictly validates a Jev request offline, without credentials. */
  validate(request: JevRequest): Promise<JevRequest> {
    return this.call('validate', { request });
  }

  /** One explicit, billed evaluation against the session budget. */
  jev(request: JevRequest, options: { timeoutMs?: number } = {}): Promise<JevAnswer> {
    return this.call('jev', { request, timeoutMs: options.timeoutMs });
  }

  // -- session ---------------------------------------------------------------

  /** Ends the session by closing its input; resolves to the exit code. */
  close(): Promise<number | null> {
    if (!this.child.stdin.destroyed) this.child.stdin.end();
    return this.exited;
  }

  // -- framing ---------------------------------------------------------------

  private call<T>(op: string, fields: Record<string, unknown>): Promise<T> {
    if (this.ended) return Promise.reject(this.ended);
    const id = this.nextId++;
    const request: Record<string, unknown> = { id, op };
    for (const [key, value] of Object.entries(fields)) {
      if (value !== undefined) request[key] = value;
    }
    return new Promise<T>((resolve, reject) => {
      this.pending.set(id, { op, resolve, reject });
      this.child.stdin.write(JSON.stringify(request) + '\n', error => {
        if (error && this.pending.delete(id)) reject(new FabricError(1, `serve input closed: ${error.message}`, op));
      });
    });
  }

  private async pump(lines: AsyncIterator<string>) {
    try {
      for (let next = await lines.next(); !next.done; next = await lines.next()) {
        const response = JSON.parse(next.value);
        const waiting = this.pending.get(response.id);
        if (!waiting) continue;
        this.pending.delete(response.id);
        if (response.ok) waiting.resolve(response.result);
        else waiting.reject(new FabricError(response.error?.code ?? 1, response.error?.message ?? 'request failed', waiting.op));
      }
    } finally {
      const code = await this.exited;
      this.ended = new FabricError(code || 1, this.log.text.trim() || `serve exited with code ${code}`);
      for (const [, waiting] of this.pending) {
        waiting.reject(new FabricError(this.ended.code, this.ended.message, waiting.op));
      }
      this.pending.clear();
    }
  }
}
